src/ic-cdk/src/api/call.rs

```diff
@@ -642,14 +651,49 @@ pub fn reply_raw(buf: &[u8]) {
     unsafe { ic0::msg_reply() };
 }
 
+#[derive(Debug)]
+/// Config to control the behavior of decoding canister endpoint arguments.
+pub struct ArgDecoderConfig {
+    /// Limit the total amount of work the deserializer can perform.
+    pub decoding_quota: Option<usize>,
+    /// Limit the total amount of work for skipping unneeded data on the wire.
+    pub skipping_quota: Option<usize>,
+    /// When set to true, print the decoding and skipping cost for each message via debug_print.
+    pub debug: bool,
+}
+
 /// Returns the argument data in the current call. Traps if the data cannot be
 /// decoded.
-pub fn arg_data<R: for<'a> ArgumentDecoder<'a>>() -> R {
+pub fn arg_data<R: for<'a> ArgumentDecoder<'a>>(arg_config: ArgDecoderConfig) -> R {
     let bytes = arg_data_raw();
 
-    match decode_args(&bytes) {
+    let mut config = DecoderConfig::new();
+    if let Some(n) = arg_config.decoding_quota {
+        config.set_decoding_quota(n);
+    }
+    if let Some(n) = arg_config.skipping_quota {
+        config.set_skipping_quota(n);
+    }
+    if arg_config.debug {
+        config.set_full_error_message(true);
```

Comment:
Does this refer to decoding costs or also full types (the latter could be too much output also in the debug mode)?

Reply:
Debug mode is supposed to be run in local replica. It not only outputs the full types, it also outputs the decoder state in case of error. Decoding cost is only shown when there is no decoding error.